File: common/src/main/java/me/pandamods/fallingtrees/client/render/TreeRenderer.java
/*
 * Copyright (C) 2024 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package me.pandamods.fallingtrees.client.render;

import com.mojang.blaze3d.vertex.PoseStack;
import me.pandamods.fallingtrees.api.TreeType;
import com.mojang.math.Quaternion;
import me.pandamods.fallingtrees.config.ClientConfig;
import me.pandamods.fallingtrees.config.FallingTreesConfig;
import me.pandamods.fallingtrees.entity.TreeEntity;
import me.pandamods.fallingtrees.utils.RenderUtils;
import me.pandamods.joml.Quaternionf;
import me.pandamods.joml.Vector3f;
import me.pandamods.joml.Math;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.entity.EntityRenderer;
import net.minecraft.client.renderer.entity.EntityRendererProvider;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.util.Mth;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.Vec3;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

@Environment(EnvType.CLIENT)
public class TreeRenderer extends EntityRenderer<TreeEntity> {
	public TreeRenderer(EntityRendererProvider.Context context) {
		super(context);
	}

	public ClientConfig getConfig() {
		return FallingTreesConfig.getClientConfig(Minecraft.getInstance().player);
	}

	@Override
	public void render(TreeEntity entity, float entityYaw, float partialTick, PoseStack poseStack, MultiBufferSource bufferSource, int packedLight) {
		TreeType tree = entity.treeType;
		if (tree == null) return;

		poseStack.pushPose();

		Map<BlockPos, BlockState> blocks = entity.getBlocks();
		float fallAnimLength = getConfig().animation.fallAnimLength;

		float bounceHeight = getConfig().animation.bounceAngleHeight;
		float bounceAnimLength = getConfig().animation.bounceAnimLength;

		float time = (float) (entity.getLifetime(partialTick) * (Math.PI / 2) / fallAnimLength);

		float fallAnim = bumpCos(time) * 90;
		float bounceAnim = bumpSin((float) ((time - Math.PI / 2) / (bounceAnimLength / (fallAnimLength * 2)))) * bounceHeight;

		float animation = (fallAnim + bounceAnim) - 90;

		Direction direction = entity.getDirection().getOpposite();
		float distance = getDistance(tree, blocks, direction.getOpposite());

		Vector3f pivot =  new Vector3f(0, 0, .5f + distance);
		pivot.rotateY(Math.toRadians(-direction.toYRot()));
		poseStack.translate(-pivot.x, 0, -pivot.z);

		Vector3f vector = new Vector3f((float) Math.toRadians(animation), 0, 0);
		vector.rotateY((float) Math.toRadians(-direction.toYRot()));
		Quaternionf quaternion = new Quaternionf().identity().rotateX(vector.x).rotateZ(vector.z);
		poseStack.mulPose(new Quaternion(quaternion.x, quaternion.y, quaternion.z, quaternion.w));

		Level level = entity.getLevel();

		poseStack.translate(pivot.x, 0, pivot.z);
		poseStack.translate(-.5, 0, -.5);
		blocks.forEach((blockPos, blockState) -> {
			poseStack.pushPose();
			poseStack.translate(blockPos.getX(), blockPos.getY(), blockPos.getZ());

			blockPos = blockPos.offset(entity.getOriginPos());
			RenderUtils.renderSingleBlock(poseStack, blockState, blockPos, level, bufferSource, packedLight);

			poseStack.popPose();
		});
		poseStack.popPose();
	}

	private float getDistance(TreeType tree, Map<BlockPos, BlockState> blocks, Direction direction) {
		float distance = 0;
		BlockPos currentPos = new BlockPos(0, 0, 0);
		BlockPos next = currentPos.relative(direction);

		while (blocks.containsKey(next)) {
			if (!tree.isTreeStem(blocks.get(next))) break;

			currentPos = next;
			next = currentPos.relative(direction);

			distance++;
		}
		BlockState blockState = blocks.get(currentPos);
		if (blockState.getOffset(Minecraft.getInstance().level, currentPos).equals(Vec3.ZERO))
			return distance - .5f;

		VoxelShape shape = blockState.getCollisionShape(Minecraft.getInstance().level, currentPos);
		if (shape.isEmpty()) shape = blockState.getShape(Minecraft.getInstance().level, currentPos);

		if (!shape.isEmpty()) {
			AABB bounds = shape.bounds();
			switch (direction) {
				case WEST -> distance -= (float) (bounds.minX);
				case EAST -> distance -= (float) (1f - bounds.maxX);
				case SOUTH -> distance -= (float) (bounds.minZ);
				case NORTH -> distance -= (float) (1f - bounds.maxZ);
			}
		} else {
			distance -= 1;
		}
		return distance;
	}

	private float bumpCos(float time) {
		return (float) Math.max(0, Math.cos(Math.clamp(-Math.PI, Math.PI, time)));
	}

	private float bumpSin(float time) {
		return (float) Math.max(0, Math.sin(Math.clamp(-Math.PI, Math.PI, time)));
	}

	@Override
	public ResourceLocation getTextureLocation(TreeEntity entity) {
		return null;
	}
}

File: src/main/kotlin/client/render/TreeRenderer.kt
/*
 * Copyright (C) 2026 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package dev.pandasystems.fallingtrees.client.render

import com.mojang.blaze3d.vertex.PoseStack
import dev.pandasystems.fallingtrees.api.TreeType
import dev.pandasystems.fallingtrees.config.fallingTreesClientConfig
import dev.pandasystems.fallingtrees.entity.TreeEntity
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.SubmitNodeCollector
import net.minecraft.client.renderer.entity.EntityRenderer
import net.minecraft.client.renderer.entity.EntityRendererProvider
import net.minecraft.client.renderer.state.CameraRenderState
import net.minecraft.client.renderer.texture.OverlayTexture
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.level.block.state.BlockState
import org.joml.Math
import org.joml.Quaternionf
import org.joml.Vector3f

class TreeRenderer(context: EntityRendererProvider.Context) : EntityRenderer<TreeEntity, TreeRenderState>(context) {
	val config get() = fallingTreesClientConfig.get()

	override fun submit(renderState: TreeRenderState, poseStack: PoseStack, nodeCollector: SubmitNodeCollector, cameraRenderState: CameraRenderState) {
		super.submit(renderState, poseStack, nodeCollector, cameraRenderState)

		val tree = renderState.treeType ?: return

		poseStack.pushPose()

		val blocks: MutableMap<BlockPos, BlockState> = renderState.blocks!!
		val fallAnimLength: Float = this.config.animation.fallAnimLength

		val bounceHeight: Float = this.config.animation.bounceAngleHeight
		val bounceAnimLength: Float = this.config.animation.bounceAnimLength

		val time = (renderState.lifeTime * (Math.PI / 2) / fallAnimLength).toFloat()

		val fallAnim = bumpCos(time) * 90
		val bounceAnim = bumpSin(((time - Math.PI / 2) / (bounceAnimLength / (fallAnimLength * 2))).toFloat()) * bounceHeight

		val animation = (fallAnim + bounceAnim) - 90

		val direction = renderState.direction!!.opposite

		val pivot = Vector3f(0f, 0f, .5f)
		pivot.rotateY(Math.toRadians(-direction.toYRot()))
		poseStack.translate(-pivot.x, 0f, -pivot.z)

		val vector = Vector3f(Math.toRadians(animation), 0f, 0f)
		vector.rotateY(Math.toRadians(-direction.toYRot()))
		val quaternion = Quaternionf().identity().rotateX(vector.x).rotateZ(vector.z)
		poseStack.mulPose(quaternion)

		poseStack.translate(pivot.x, 0f, pivot.z)
		poseStack.translate(-.5, 0.0, -.5)
		blocks.forEach { (blockPos, blockState) ->
			poseStack.pushPose()
			poseStack.translate(blockPos.x.toFloat(), blockPos.y.toFloat(), blockPos.z.toFloat())
			nodeCollector.submitBlock(poseStack, blockState, renderState.lightCoords, OverlayTexture.NO_OVERLAY, renderState.outlineColor)
			poseStack.popPose()
		}
		poseStack.popPose()
	}

	override fun createRenderState(): TreeRenderState {
		return TreeRenderState()
	}

	override fun extractRenderState(entity: TreeEntity, renderState: TreeRenderState, f: Float) {
		renderState.treeType = entity.treeType
		renderState.blocks = entity.blocks
		renderState.lifeTime = entity.getLifetime(f).toDouble()
		renderState.direction = entity.direction
		renderState.originPos = entity.originPos
		renderState.level = entity.level()
		super.extractRenderState(entity, renderState, f)
	}

	private fun bumpCos(time: Float): Float {
		return Math.max(0.0, Math.cos(Math.clamp(-Math.PI, Math.PI, time.toDouble()))).toFloat()
	}

	private fun bumpSin(time: Float): Float {
		return Math.max(0.0, Math.sin(Math.clamp(-Math.PI, Math.PI, time.toDouble()))).toFloat()
	}

	override fun affectedByCulling(display: TreeEntity): Boolean {
		return false
	}
}
